src/hooks/resources/device.ts

```diff
@@ -579,3 +595,105 @@ async function checkSupervisorReleaseUpgrades(
 		}
 	}
 }
+
+async function setSupervisorReleaseResource(
+	api: sbvrUtils.PinejsClient,
+	deviceIds: number[],
+	supervisorVersion: string,
+) {
+	if (deviceIds.length === 0) {
+		return;
+	}
+	const devices = (await api.get({
+		resource: 'device',
+		options: {
+			$expand: {
+				is_of__device_type: {
+					$select: 'id',
+				},
+			},
+			// if the device already has a supervisor_version, just bail.
+			$filter: {
+				id: { $in: deviceIds },
+				supervisor_version: { $eq: null },
+			},
+			$select: ['id', 'supervisor_version'],
+		},
+	})) as AnyObject;
+
+	if (devices.length === 0) {
+		return;
+	}
+
+	const deviceTypes = _.uniqBy(
+		_.reduce(
+			devices,
+			(result: AnyObject[], val, _key) => {
+				// TODO: hardcoded index
+				result.push(val.is_of__device_type[0].id);
+				return result;
+			},
+			[],
+		),
+		'id',
+	);
+
+	if (deviceTypes.length === 0) {
+		return;
+	}
+
+	await Bluebird.map(deviceTypes, async (deviceType) => {
+		const affectedDeviceIds = _.map(devices, (d) => {
+			// TODO: hardcoded index
+			if (d.is_of__device_type[0].id === deviceType) {
+				return d.id;
+			}
+		});
+		if (affectedDeviceIds.length === 0) {
+			return;
+		}
+
+		const rootApi = api.clone({
+			passthrough: {
+				req: root,
+			},
+		});
+		const supervisorRelease = (await rootApi.get({
+			resource: 'supervisor_release',
+			options: {
+				$select: ['id'],
+				$filter: {
+					supervisor_version: `v${supervisorVersion}`,
+					is_public: true,
+					is_for__device_type: {
+						$any: {
+							$alias: 'dt',
+							$expr: {
+								dt: {
+									id: deviceType,
+								},
+							},
+						},
+					},
+				},
+			},
+		})) as AnyObject[];
+
+		if (supervisorRelease.length === 0) {
+			return;
+		}
+
+		await rootApi.patch({
+			resource: 'device',
+			options: {
+				$filter: {
+					id: { $in: affectedDeviceIds },
+				},
+			},
+			body: {
+				// TODO: hardcoded index
```

Comment:
Discussed offline, since this is defined in SBVR as `Necessity: each device is of exactly one device type.` (https://github.com/balena-io/open-balena-api/blob/master/src/balena.sbvr#L433), we can comfortably ignore this (or further clean it up by using the `__id` selector).

Reply:
included in latest commit.